### ISSUE TITLE
Bug: Pasting nested lists doesn't work which copied from word 

### ISSUE THREAD

**Issue by pradeep-jai** [facebook/lexical#4922]
Not getting the same style of the content when I copy and paste it from word to lexical text editor. How we can do that ?

Eg : 

1. This is format in word  with indent:

        • One
        • Two
        • Three
              o 3.1
              o 3.2

2. But the format comes from lexical after the pasting the same format which copied from word

•	One
•	Two
•	Three
o	3.1
o	3.2

Lexical version: latest

## Steps To Reproduce

1. Get the some nested list value from word documents same like above
2. Paste it into lexical editor

I have attached the video. Please have check on it.

[Format_issue_lexical.webm](https://github.com/facebook/lexical/assets/52864573/f30caa6a-bff6-4699-8206-e623fea7d42f)
[Format_issue_lexical.webm](https://github.com/facebook/lexical/assets/52864573/ba4a835d-4b1a-40fd-8412-7142fe457326)

**Comment by smrifat1411** [comment]
This isn't really Lexical losing the formatting — it's that Word never sent a list in the first place.

When you copy a list out of Word, there is no `<ul>` and no `<li>` on the clipboard. Every item arrives as a paragraph:

```html
<p class=MsoListParagraphCxSpFirst style='mso-list:l0 level1 lfo1'>
  <![if !supportLists]>
    <span style='mso-list:Ignore'>o<span style='font:7.0pt'>&nbsp;&nbsp;</span></span>
  <![endif]>
  3.1
</p>
```

Three things to notice:

- `mso-list:l0 level1 lfo1` is the **only** record that this paragraph belongs to a list. `l0` says which list, `level1` says how deep — that's your nesting.
- The bullet is **literal text**, not a rendered marker. That's why you see `o` in your output. It isn't the letter o either — it's a hollow bullet Word draws in Courier New, pasted as a character.
- The span holding it is labelled `mso-list:Ignore`. Word is telling you it's presentation, not content.

`$generateNodesFromDOM` is doing the right thing with what it's given: it sees paragraphs, so it makes paragraphs. To get a list back you have to rebuild it from those `mso-list` markers *before* the HTML reaches Lexical — and critically, before you strip the Word styling, because stripping it destroys the only evidence of the structure.

I hit this building an exam authoring tool and ended up writing it as a small library — [wordpaste](https://github.com/smrifat1411/wordpaste) (MIT, zero dependencies, ~3.5 kB). Disclosure: I'm the author. It groups the paragraphs by list id, nests them by level, reads each marker to decide ordered vs bulleted, and drops the marker spans:

```js
import { PASTE_COMMAND, COMMAND_PRIORITY_HIGH, $getRoot, $insertNodes } from 'lexical';
import { $generateNodesFromDOM } from '@lexical/html';
import { transformPastedHTML } from 'wordpaste';

editor.registerCommand(
  PASTE_COMMAND,
  (event) => {
    const html = event.clipboardData?.getData('text/html');
    if (!html) return false;

    event.preventDefault();
    const dom = new DOMParser().parseFromString(transformPastedHTML(html), 'text/html');

    editor.update(() => {
      $getRoot().selectEnd();
      $insertNodes($generateNodesFromDOM(editor, dom));
    });
    return true;
  },
  COMMAND_PRIORITY_HIGH,
);
```

Your exact case comes out as:

```html
<ul>
  <li>One</li>
  <li>Two</li>
  <li>Three
    <ul><li>3.1</li><li>3.2</li></ul>
  </li>
</ul>
```

It also recovers Word equations as editable LaTeX rather than screenshots, which is the other thing that silently disappears on paste.

If you'd rather not add a dependency, the approach above is the part that matters — read the `mso-list` markers first, rebuild, *then* clean. You can [paste a document into the playground](https://smrifat1411.github.io/wordpaste/) to see what your own clipboard actually contains.

**Comment by smrifat1411** [comment]
The list half of this is already solved in-tree, it just isn't shipped.

`dev-examples/dom-import/src/wordPaste.ts` builds real nested `ol`/`ul` out of Word's
`MsoListParagraph` runs (snapshotting `mso-list` to `data-mso-list` before the
stylesheet-inlining preprocess drops it), and
`packages/lexical-list/src/__tests__/unit/ListImportExtension.test.ts` carries its own
copy of that logic. But `@lexical/list`'s shipped `ListImportExtension` has no `mso`
handling, so a plain paste from Word still arrives as flat paragraphs.

#8528 also set the precedent for where the fix belongs: VS Code paste ships as
`$installVscodeCodePasteOverlay` from `@lexical/code-core`, while Word paste stayed
in examples. I'd like to give Word lists the same treatment — export an
`$installWordListPasteOverlay` (`DOMPreprocessFn`) from `@lexical/list`, wire it
through `ListExtension`'s `DOMImportExtension` config, and have both the example and
the test import it instead of duplicating it. Still gated on the
`<meta name="Generator" content="Microsoft Word…">` sniff, so non-Word pastes pay
nothing.

@etrepum you built this pipeline — happy to open that PR if that's the shape you
want, or to leave it in examples and document it instead if keeping it out of the
package was deliberate.

Context on why I'm offering: I built and maintain
[wordpaste](https://github.com/smrifat1411/wordpaste), which handles this same Word
clipboard shape standalone. Equations are the other half — Word ships them as OMML
and nothing here converts them yet — but that's a separate PR if there's appetite.

**Comment by etrepum** [comment]
I don't think that OMML belongs in a published package because there is no published EquationNode for it to interop with. That's probably best maintained out of tree, or at best in an example, there are a lot of choices for how you might want to deal with equations.

I don't have a strong opinion about whether word lists are specifically handled in ListImportExtension but perhaps a suite of office importers is better, e.g. @lexical/ms-office or something like that which could have opt-in code for handling word/excel/etc. interop concerns. Could also be something out of tree.

**Comment by smrifat1411** [comment]
Fair on OMML — dropping it. There's no published node to import into and the right target differs per app, so that belongs out of tree.

On lists, one detail that might make the call easier: the logic is already in the repo twice, and the two copies have drifted.

`dev-examples/dom-import/src/wordPaste.ts` and `packages/lexical-list/src/__tests__/unit/ListImportExtension.test.ts` carry the same functions under the same names — `readWordListLevel`, `$findMarkerSpan`, `$readWordMarker`, `classifyWordListType`, `$stripWordMarker`, `$buildWordListTree`, `$installWordOverlay`. The example's `$installWordOverlay` snapshots `mso-list` onto `data-mso-list` first, because `$inlineStylesFromStyleSheets` re-serializes the style attribute and drops unknown properties. The test's copy has no snapshot and reads the style attribute directly — so the drop that stash exists to survive is the one thing not covered.

That's the cost today regardless of where the code lands: two implementations, one of them tested, neither shipped.

`@lexical/ms-office` seems right if the scope grows — Word tables, Excel — but there'd be exactly one importer in it today, so I'd rather not stand up a package on spec.

Smallest version I'd propose: export the preprocess from `@lexical/list` as `$installWordListPasteOverlay`, opt-in through `ListExtension`'s `DOMImportExtension` config and off unless configured, with the example and the test importing it. If `@lexical/ms-office` happens later, that's a re-export rather than a rewrite.

If you'd rather it stay in examples, I'll send a PR that only dedupes the two copies and leaves the package surface alone. Say which and I'll open it.

**Comment by etrepum** [comment]
If it can be opt-in and tree-shakeable in @lexical/list that should be fine.

**Comment by smrifat1411** [comment]
Implemented, in the shape you approved.

`@lexical/list` now exports `$installWordListPasteOverlay` — a `DOMPreprocessFn`, not an extension. `ListExtension` doesn't reference it, so it costs nothing unless an app opts in through the `DOMImportExtension` config:

```ts
defineExtension({
  dependencies: [
    ListExtension,
    configExtension(DOMImportExtension, {
      preprocess: [$installWordListPasteOverlay],
    }),
  ],
  name: 'my-editor',
});
```

The rules, the overlay and the import state are module-scope `/* @__PURE__ */` calls in their own module, so a bundle that never names the export drops all of it.

It replaces the two copies already in the repo — `dev-examples/dom-import/src/wordPaste.ts` and the inline copy in `ListImportExtension.test.ts`. Both deleted; the example and the test now use the shipped export. Net −112 lines.

One behavioural detail, since the two copies disagreed and the shipped one keeps the example's behaviour: Word's real clipboard payload carries a `<style>` block, so the default `$inlineStylesFromStyleSheets` preprocess rewrites each matched element's style attribute through `CSSStyleDeclaration` and silently drops `mso-list`, which is non-standard. The overlay's preprocess runs first and snapshots it onto `data-mso-list` so levels and markers still resolve. The test copy had no such snapshot and its fixture had no `<style>` block, so nothing caught it — there's now a test using realistic Word CSS that collapses the nesting without the snapshot and passes with it.

The overlay also carries the `<o:p>` rule (Office's paragraph-end marker), since Word emits one inside every list paragraph. Happy to drop it if you'd rather that not live in `@lexical/list`.

PR: #9082
